### PR TITLE
Fix buildscripts for iOS to bring back iOS support in the future

### PR DESCRIPTION
**Pull request by lmharpae**
As an EasyRPG user who has some interest at developing EasyRPG for iOS, I want to fix the current buildscripts for iOS that were unable to build before.

### I already tested the toolchain building using the computer below:

* **Computer used:** MacBook Pro 13-inch (Late 2013)
* **OS:** macOS Catalina (10.15.7)
* **Xcode version:** 12.4
* **iOS SDK version:** 14.4
* **Apple Clang version:** 12.0
* **CMake version:** 3.27.8

The fixed buildscripts will generate toolchain for `armv7`, `arm64` and `universal`, seperated into 3 respective folders. I already compressed the `universal` one into this file below containing the latest version of all libraries for iOS:

[ios_toolchain.tar.gz](https://drive.google.com/file/d/1a0Q9A0ox79UkukF6vraeqkGjDy92sX8T/view) (70.0 MB)

I also tested building the latest version of `liblcf` with the toolchain above:

[liblcf-ios.tar.gz](https://drive.google.com/file/d/1Jv_KhqQB_f3P7zYCSCPHeIwuqbpJx6vP/view) (34.8 MB)

However, while trying to build the latest version of `Player`, I encountered an error that some codes in the Player only supports iOS 11 or later, but it seemed that the toolchain could be used without any errors.

### CODE REVIEW

File: ios/2_build_toolchain.sh [hunk 1]
```diff
@@ -77,31 +87,38 @@ function install_lib_sdl2() {
...
+	install_lib $LHASA_DIR $LHASA_ARGS
+	install_lib_cmake $FMT_DIR $FMT_ARGS
+	install_lib_icu_cross
+	install_lib_liblcf
+	install_lib_sdl2
+}
+
+# Emptying harfbuzz and freetype's build arguments because they aren't supported
+HARFBUZZ_ARGS=""
+FREETYPE_ARGS=""
```
Comment (Ghabry) [inline ios/2_build_toolchain.sh:119]: The reason why the freetype and harfbuzz arguments do not work is that you forgot to add ``_cmake`` to ``install_lib``

File: ios/2_build_toolchain.sh [hunk 2]
```diff
@@ -77,31 +87,38 @@ function install_lib_sdl2() {
...
-install_lib_cmake $FLUIDLITE_DIR $FLUIDLITE_ARGS -DENABLE_SF3=ON
-install_lib_meson $INIH_DIR $INIH_ARGS
-install_lib $LHASA_DIR $LHASA_ARGS
-install_lib_cmake $FMT_DIR $FMT_ARGS
-install_lib_icu_cross
-install_lib_liblcf
-
-install_lib_sdl2
+function build() {
+	install_lib_icu_native
```
Comment (Ghabry) [inline ios/2_build_toolchain.sh:91]: Please move this icu native build out (like macOS does it). The native ICU must be built for the platform you are building on (macOS), not the target (iOS).

File: ios/2_build_toolchain.sh (hunk 1 above)
Comment (lmharpae) [reply ios/2_build_toolchain.sh:119]: > The reason why the freetype and harfbuzz arguments do not work is that you forgot to add `_cmake` to `install_lib`

So I have to use CMake to install Freetype and Harfbuzz, not Make and Autotools?

**Comment by Ghabry** [comment]
sorry for not giving more feedback on this. All my recent attempts to boot the latest macOS in a VM (KVM linux) failed with a kernel panic. :/

**Comment by lmharpae** [comment]
> sorry for not giving more feedback on this. All my recent attempts to boot the latest macOS in a VM (KVM linux) failed with a kernel panic. :/

As I know, there's currently **no way** to boot the latest version of macOS on virtual machines. I also tried booting macOS 13 (Ventura) before but it didn't succeed.

However, I think the older version of macOS is enough; remember that I edited and built the buildscripts using macOS 10.15 (Catalina), so can you try booting Catalina or Big Sur on VM?

Btw, I also bought a new MacBook Pro 2018 some days before (cause my MBP Late 2013 mentioned in this PR was already broken), so let me know if you need any help.

**Comment by Ghabry** [comment]
ah okay, this explains it. I will try macOS 12 when I find time and report back. Thanks.

**Comment by Ghabry** [comment]
Uhm yesterday I saw a comment in the configuration file that the CPU must be changed from ``Penryn`` to ``Haswell-noTSX`` for Sonoma and... my installation I made a month ago booted.

The background image is not rendering (desktop is white) and audio is broken but who cares.

grafik

**Comment by lmharpae** [comment]
> Uhm yesterday I saw a comment in the configuration file that the CPU must be changed from ``Penryn`` to ``Haswell-noTSX`` for Sonoma and... my installation I made a month ago booted.
> 
> So obvious case of RTFM xD
> 
> The background image is not rendering (desktop is white) and audio is broken but who cares.
> 
> grafik
> 

"The background image is not rendering" and especially the "Display 7 MB" at the Graphics part means there's no graphics acceleration at all, so the UI and graphics-related experience will be very bad.

But since we only use it for building the scripts, I think the graphics acceleration is not necessary.

> I used to try macOS on VM before, both on VMware and VirtualBox; and I don't know if that was because of my misconfiguration or anything else, but on my computer the macOS performs better on VirtualBox than VMware. But that was macOS Catalina, and I don't know if that also applies to macOS Sonoma, as I already got a MacBook before Sonoma was even released.
> But yes, there is **no way** to enable graphics acceleration on virtual machines, especially when no VM software _officially_ support installing macOS.
>
> Finally, if you can consider it; why don't you try to Hackintosh?

**Comment by Ghabry** [comment]
It compiled for me after bumping the iphoneos version to "9.0".

Thanks for that contribution. 

I will rebase and merge it when #167 is in.

Then it will be become part of our automatic library update system and shouldn't break anytime soon.

---

yeah I know that it has no hardware acceleration because its just running in KVM on Linux. But it feels faster than my Windows 11 VM.

For testing whether it compiles and runs it is good enough.

> Finally, if you can consider it; why don't you try to Hackintosh your PC? That is 
much better than installing on a VM.

I'm not a fan of macOS. I only use it when testing Player development.

**Comment by carstene1ns** [comment]
Is the SDL2 IOS patch upstreamed yet?

**Comment by Ghabry** [comment]
My guess here is that the wrong buildscript for iOS is invoked (the patch looks like replacing macOS headers with iOS headers).

**Comment by fdelapena** [comment]
> > Is the SDL2 IOS patch upstreamed yet?
> 
> I don't fully understand what does the "upstreamed" word mean in this situation.

It seems it's about the patch replacing Cocoa with Foundation headers. If it's the solution for everybody using SDL for iOS, then it is worth to be sent to SDL auhors to fix the build there (upstream) instead of keeping that particular patch here or in distributions (downstream).
https://en.m.wikipedia.org/wiki/Upstream_(software_development)

**Comment by Ghabry** [comment]
Don't worry about it. We will fix/ check this before merging.

**Comment by Ghabry** [comment]
To convince SDL2 that we build for iOS the triplet must contain "-iOS-". Though then I get later build errors because of ``-Werror,-Wdeclaration-after-statement``.

With CMake it builds but only because they didn't enable that ``-Werror`` for it. :sweat_smile: (for SDL3 they discuss to just deprecate C89 which would get rid of that warning).

**Comment by Ghabry** [comment]
Thanks again. I rebased it and made some small changes in #173 . Closing this one.

**Comment by Ghabry** [comment]
The buildsystem is now updated to macOS Sequoia. Which means (library) building for iOS will come soon :). Player is a different topic...

**Comment by Serena1432** [comment]
> The buildsystem is now updated to macOS Sequoia. Which means (library) building for iOS will come soon :). Player is a different topic...

I've successfully built the libretro version for iOS before, so I think I can contribute on that one. Too bad I don't know much about iOS development (yet), so I don't know how should I contribute for the Player.

**Pull request closed without merging**